### PR TITLE
adding mixed precision training with Apex

### PR DESCRIPTION
**Pull request by vinhngx**
- Adding mixed precision training for 2-3x more throughput on single GPU with tensor cores. 
- Adding an additional metric to monitor throughput (imgs/s)

### CODE REVIEW

File: references/classification/train.py [hunk 1]
```diff
@@ -11,26 +11,38 @@
...
+                scaled_loss.backward()
+        else:
+            loss.backward()
         optimizer.step()
 
         acc1, acc5 = utils.accuracy(output, target, topk=(1, 5))
         batch_size = image.shape[0]
         metric_logger.update(loss=loss.item(), lr=optimizer.param_groups[0]["lr"])
         metric_logger.meters['acc1'].update(acc1.item(), n=batch_size)
-        metric_logger.meters['acc5'].update(acc5.item(), n=batch_size)
```
Comment (fmassa) [inline references/classification/train.py]: Why did you remove `acc5`?

File: references/classification/train.py [hunk 2]
```diff
@@ -11,26 +11,38 @@
...
     metric_logger.add_meter('lr', utils.SmoothedValue(window_size=1, fmt='{value}'))
+    metric_logger.add_meter('img/s', utils.SmoothedValue(window_size=10, fmt='{value}'))
+
     header = 'Epoch: [{}]'.format(epoch)
     for image, target in metric_logger.log_every(data_loader, print_freq, header):
         image, target = image.to(device), target.to(device)
         output = model(image)
         loss = criterion(output, target)
 
+        start_time = time.time()
```
Comment (fmassa) [inline references/classification/train.py]: Is this necessary, given that we already log timing inside `metric_logger` for the full forward / backward / optimizer?

File: references/classification/train.py [hunk 3]
```diff
@@ -11,26 +11,38 @@
 
 import utils
 
+try:
```
Comment (fmassa) [inline references/classification/train.py]: Can we move this import later on, only if the user pass the `--apex` flag?

File: references/classification/train.py (hunk 1 above)
Comment (vinhngx) [reply references/classification/train.py]: good catch :). This is an accidental removal. I re-added it.

File: references/classification/train.py (hunk 2 above)
Comment (vinhngx) [reply references/classification/train.py]: maybe not, but I just want to have a simple and intuitive way to update the throughput instead of pulling it from `metric_logger`. Appreciate any suggestion though.

File: references/classification/train.py (hunk 3 above)
Comment (vinhngx) [reply references/classification/train.py]: the import should be at module level. I'm not sure how we can check the `--apex` flag before entering `main`.

**Comment by vinhngx** [comment]
> Also, I believe APEX is Python3-only, right?

Indeed, as stated in `https://github.com/NVIDIA/apex`, python3 is required.

File: references/classification/train.py (hunk 2 above)
Comment (fmassa) [reply references/classification/train.py]: The time for the total iteration gets printed every `print_freq` https://github.com/pytorch/vision/blob/47db963800649bdd609d619bbc5f08052fe5520a/references/classification/utils.py#L123

So I suppose this is enough and we can just remove this extra timing?

File: references/classification/train.py (hunk 3 above)
Comment (fmassa) [reply references/classification/train.py]: In this case, I'd prefer to do it slightly differently.

Currently, we are unconditionally printing `Please install apex from https://www.github.com/nvidia/apex to enable mixed-precision training.`, even if the user has not requested it.

What about the following:

```python
try:
    from apex import amp, optimizers
except ImportError:
    amp, optimizers = None, None
```

and then, later on in the code, if the user passes `--apex`, we check that both `amp` and `optimizers` is not `None`, and if that's not the case raise an error?

File: references/classification/train.py [hunk 4]
```diff
@@ -11,26 +11,39 @@
...
+    from apex import amp, optimizers
```
Comment (fmassa) [inline references/classification/train.py]: do we use `optimizers` anywhere in the code?

File: references/classification/train.py (hunk 2 above)
Comment (vinhngx) [reply references/classification/train.py]: I guess most people cannot easily relate to a time statistics (sec per batch) compared to a throughput statistics (imgs/sec), hence the reason for adding a throughput metric. If removing the added timing, do you have any suggestion to log the throughput?

File: references/classification/train.py (hunk 3 above)
Comment (vinhngx) [reply references/classification/train.py]: perfect. Let's go with this.

Comment (vinhngx) [reply references/classification/train.py]: currently not using "optimizers". Removing.

File: references/classification/train.py [hunk 5]
```diff
@@ -68,6 +82,14 @@ def _get_cache_path(filepath):
 
 
 def main(args):
+    if args.apex:
+        if (sys.version_info < (3, 0)):
+            print("Apex currently only supports Python 3. Aborting.")
```
Comment (fmassa) [inline references/classification/train.py]: nit: you can raise an error in this case
```python
raise RuntimeError("Apex currently only supports Python 3. Aborting.")
```

File: references/classification/train.py [hunk 6]
```diff
@@ -68,6 +82,14 @@ def _get_cache_path(filepath):
...
+            return
+        if amp is None:
+            print("Please install apex from https://www.github.com/nvidia/apex to enable mixed-precision training.")
```
Comment (fmassa) [inline references/classification/train.py]: Similarly here, we could raise a `RuntimeError`

File: references/classification/train.py (hunk 2 above)
Comment (fmassa) [reply references/classification/train.py]: Let's keep this then, even though this is a bit redundant IMO.

File: references/classification/train.py (hunk 5 above)
Comment (vinhngx) [reply references/classification/train.py]: yep very good suggestion. I've edited this.

File: references/classification/train.py (hunk 6 above)
Comment (vinhngx) [reply references/classification/train.py]: good suggestion. I've edited this.

**Comment by mats-claassen** [comment]
Apex has been supported in `classification/train.py` but not in `detection/engine.py` for example. Is there a reason for that or could Apex be supported for object detection as well?

**Comment by fmassa** [comment]
@mats-claassen I think making detection training support apex is going to be a bit more involved, so I think for now I would rather not do it

**Comment by vinhngx** [comment]
Thanks Mathias. Indeed we've been working on incorporating Apex mixed
precision training to detection models. Last time I checked there were a
few unsupported ops, but if you want to have a try with the latest Apex
(now incorporated into Pytorch) pls go ahead :)
